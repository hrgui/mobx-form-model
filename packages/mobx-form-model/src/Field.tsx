
import * as React from 'react';
import { observer, inject } from 'mobx-react';
import { getIn, isFunction, isString } from './utils';
import filterReactProps from 'filter-react-props';
import { toJS, isObservable } from 'mobx';

@inject((stores: any) => {
  return {model: stores.model};
})
@observer
export default class Field extends React.Component<any, any> {
  render() {
    const {
      type,
      name,
      value,
      component = 'input',
      render,
      model
    } = this.props;

    if (!model) {
      return null;
    }

    let modelValue = getIn(model.values, name);
    const v = (type === 'radio' || type === 'checkbox'
          ? value
          // why OR '' so we dont get uncontrolled error
          : modelValue) || '';
    const defaultProps: any = {
      defaultValue: v,
      value:v,
      onChange: model.handleChange,
      onBlur: model.handleBlur
    };

    if (type === 'radio') {
      defaultProps.checked = modelValue === this.props.value;
    }

    const props = {
      ...defaultProps,
      ...this.props
    };

    if (isObservable(props.value) && !this.props.bindValueToMobx) {
      props.value = toJS(props.value);
    }

    if (isFunction(render)) {
      return render(props);
    }

    let finalProps = isString(component)
      ? filterReactProps(props)
      : props;

    if (!isString(component)) {
      delete finalProps.component;
      delete finalProps.model;
      delete finalProps.defaultValue;
    }

    return React.createElement(component, {
      ...finalProps
    });
  }
}
