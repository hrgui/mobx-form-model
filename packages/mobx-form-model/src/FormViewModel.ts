import { observable, set as mobXSet } from 'mobx';
import {
  isFunction,
  setNestedObjectValues,
  setIn,
  yupToFormErrors,
  validateYupSchema,
  getIn
} from './utils';
import { set } from 'lodash';

export default interface FormViewModel {
  /**
   * This method runs on every form update.
   * If this returns an object that has values, the form will be in an invalid state.
   */
  validate?: ((values) => any);
  /**
   * This method runs on every form update as well.
   * The difference here is that this is a validationSchema that is provided by yup.
   * It can be either a function or not.
   */
  validationSchema: any | (() => any);
  /**
   * When the user presses the cancel button, currently it will do nothing.
   */
  onCancel?: (() => any);
};

export default class FormViewModel {
  /**
   * This is the initial state of the form. This gets copied over to `values` in constructor runtime. On reset, this is used to reset the values to an initial state.
   */
  initialValues: any = {};
  /**
   * These are the values of the form - the current state. Note that unlike formik, the values change automatically instead of waiting on the next render cycle - so there's a huge difference there.
   */
  @observable values: any = { ...this.initialValues };
  /**
   * This is used when any of the fields mentioned are touched; meaning interacted with. By default, `<Field />` uses onBlur for this event.
   */
  @observable touched: any = {};
  /**
   * This is an object that consists of all the errors.
   */
  @observable errors: any = {};
  /**
   * This is true when the form is currently submitting, false otherwise.
   */
  @observable isSubmitting = false;
  /**
   * This is to keep track of how many times the user clicked on the submit.
   */
  @observable submitCount = 0;
  /**
   * If provided, after all validations are run for the parent, it will run validation for all children.
   * Note: This only can handle children at the root level. If desired to handle children at a level other
   * than root, please do so at the FormViewModel.
   */
  @observable childFormModels: {[key: string]: FormViewModel} = {};
  /**
   * If provided, when setting the field as touched, it will inform the parent about the field being touched.
   * Note: This only can handle children at the root level. If desired to handle children at a level other
   * than root, please do so at the FormViewModel.
   */
  @observable childProp;
  parentViewModel;

  constructor(values?) {
    this.onReset = this.onReset.bind(this);
    this.setErrors = this.setErrors.bind(this);
    this.setFieldError = this.setFieldError.bind(this);
    this.setFieldTouched = this.setFieldTouched.bind(this);
    this.setValues = this.setValues.bind(this);
    this.setFieldValue = this.setFieldValue.bind(this);
    this.handleBlur = this.handleBlur.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleCancel = this.handleCancel.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);

    if (values) {
      this.setInitialValues(values);
      this.setValues(values, false);
    }
  }

  /**
   * When the user has succesfully saved the form
   * @param response
   */
  onSubmitSuccess(response) {
    return response;
  }
  /**
   * When the user has passed validation and is now going to submit.
   */
  onSubmit(values?) {
    return values;
  }

  /**
   * Determines if the form is valid.
   */
  get isValid() {
    return Object.keys(this.errors).length === 0;
  }

  /**
   * This is a generic method to get the value from inputEvent.
   * The code tries to pass in e.target into target, but if target doesn't exist, target is the event
   * property.
   *
   * Now if neither doesn't work, use other to figure out what should be returned as a value to the model.
   *
   * @param target
   * @param other
   */
  getValueFromInputEvent(target, ...other) {
    const { type, value, checked } = target;
    const isNumericalValue = /number|range/.test(type);
    const isCheckbox = /checkbox/.test(type);

    if (isNumericalValue) {
      let parsed = parseFloat(value);
      return isNaN(parsed) ? '' : parsed;
    }

    if (isCheckbox) {
      return checked;
    }

    return value;
  }

  /**
   * This gets the name to be applied to the values property.
   * By default, it will assume name as the property, but if id is present it will use that too.
   * @param target
   * @param other
   */
  getNameFromInputEvent(target, ...other) {
    const { name, id } = target;
    return name || id;
  }

  /**
   * This method is the default handler for handling change events.
   */
  handleChange(e, ...other) {
    const target = e.target || e;
    const fieldName = this.getNameFromInputEvent(target, ...other);
    const fieldValue = this.getValueFromInputEvent(target, ...other);
    if (e.persist) {
      e.persist();
    }
    this.setFieldValue(fieldName, fieldValue);
  }

  /**
   * This method is the default handler for handling blur events.
   */
  handleBlur(e, ...other) {
    const target = e.target || e;
    const fieldName = this.getNameFromInputEvent(target, ...other);
    this.setFieldTouched(fieldName, true);
  }

  /**
   * This method sets a field value explicitly.
   * The added benefit is that validations will be run after setFieldValue is called.
   * Note that setFieldValue will automatically do the change immediately; there is no need to wait for another cycle.
   */
  setFieldValue(name, value) {
    this.setInMobx('values', name, value);
    this.runValidations();
  }

  setInitialValues(values) {
    // we make a copy so that if we need it, we can refer to it
    this.initialValues = {...values};
  }

  /**
   * This method sets all field values explicitly.
   * Note that this will cause every form value to re-render, because the entire values object has changed.
   * The added benefit is that validations will be run after setValues is called.
   * Note that setValues will automatically do the change immediately; there is no need to wait for another cycle.
   */
  setValues(values, validate=true) {
    this.values = values;
    validate && this.runValidations();
  }

  /**
   * This method sets a single field touched value explicitly.
   * The added benefit is that validations will be run after setFieldTouched is called.
   * Note that setFieldTouched will automatically do the change immediately; there is no need to wait for another cycle.
   */
  setFieldTouched(name, value) {
    this.setInMobx('touched', name, value);

    if (this.parentViewModel) {
      this.parentViewModel.setFieldTouched(`${this.childProp}.${name}`, value);
    }

    this.runValidations();
  }

  /**
   * This method sets a single field error value explicitly.
   * Note that setFieldError will automatically do the change immediately; there is no need to wait for another cycle.
   */
  setFieldError(name, value) {
    this.setInMobx('errors', name, value);
  }

  /**
   * This method changes all the errors explicilty.
   * Note that setErrors will automatically do all the changes immediately, no need to wait for another cycle.
   */
  setErrors(errors) {
    this.errors = errors;
  }

  /**
   * Sets the form as currently being submitted or not.
   * @param value
   */
  setSubmitting(value) {
    this.isSubmitting = value;
  }

  addChildFormModel(fm: FormViewModel, key) {
    fm.childProp = key;
    fm.parentViewModel = this;
    mobXSet(this.childFormModels, {[key]: fm});
  }

  /**
   * This method is the core method for all the validations to run.
   * It returns a boolean: if its true then the form is valid, false otherwise.
   * @param values
   */
  async runValidations(values = this.values): Promise<boolean> {
    let schemaErrors;
    let validateFnErrors;

    if (this.validationSchema) {
      schemaErrors = await this.runValidationSchema(values);
    }

    if (this.validate) {
      validateFnErrors = await this.validate(this.values);
    }

    this.errors = { ...schemaErrors, ...validateFnErrors };

    if (this.childFormModels) {
      for (const name in this.childFormModels) {
        if (!this.childFormModels.hasOwnProperty(name)) {
          continue;
        }

        const formViewModel = this.childFormModels[name];

        const isValid = await formViewModel.runValidations();
        if (isValid) {
          continue;
        }

        this.errors = {...this.errors, ...{[formViewModel.childProp]: {...formViewModel.errors}} };
      }
    }

    return Object.keys(this.errors).length === 0;
  }

  /**
   * By default, this resets the form to a pristine state. If this method is overridden, please call `super.onReset()` to get to an initial state.
   */
  onReset() {
    this.setValues(this.initialValues, false);
    this.touched = {};
    this.errors = {};
    this.isSubmitting = false;
    this.submitCount = 0;
  }

  /**
   * This method handles the submit event.
   * @param e
   */
  handleSubmit(e) {
    if (e && e.preventDefault) {
      e.preventDefault();
    }

    return this.submitForm();
  }

  /**
   * This method handles the reset event.
   */
  handleReset() {
    if (this.onReset) {
      this.onReset();
    }
  }

  /**
   * This method handles the cancel event.
   */
  handleCancel() {
    if (this.onCancel) {
      this.onCancel();
    }
  }

  /**
   * This method is the main submitForm handler when the submit event is invoked.
   * 1. We set all form values as touched
   * 2. We set isSubmitting = true then set it to false afterward.
   */
  async submitForm() {
    try {
      this.touched = setNestedObjectValues(this.values, true);
      this.isSubmitting = true;
      this.submitCount = this.submitCount + 1;

      const isValid = await this.runValidations();

      if (!isValid) {
        throw this.errors;
      }

      const response = await this.onSubmit(this.values);
      if (this.onSubmitSuccess) {
        await this.onSubmitSuccess(response);
      }
      this.isSubmitting = false;
    } catch (e) {
      this.isSubmitting = false;
      return this.onSubmitError(e);
    }
  }

  /**
   * When the user submits but an error occurs, by default it will just throw a console.error.
   * @param e
   */
  onSubmitError(e) {
    console.error('Error when submitting form', e);
  }

  /**
   * This runs the validation schema and translates it to a more readable form that mobx-form-model can present.
   * @param values
   */
  async runValidationSchema(values) {
    const schema = isFunction(this.validationSchema)
      ? this.validationSchema()
      : this.validationSchema;
    try {
      await validateYupSchema(values, schema);
      return {};
    } catch (err) {
      const errors = yupToFormErrors(err);
      return errors;
    }
  }

  /**
   * This method sets a value in any object here w/ mobx. It uses set if the value is defined, it overwrites the entire thing
   * otherwise.
   * @param prop
   * @param name
   * @param value
   */
  private setInMobx(prop, name, value) {
    const lastValue = getIn(this[prop], name);

    if (lastValue === undefined) {
      this[prop] = setIn(this[prop], name, value);
      return;
    }

    return set(this[prop], name, value);
  }
}
