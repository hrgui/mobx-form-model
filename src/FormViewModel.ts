import {observable} from 'mobx';
import {isFunction, setNestedObjectValues, setIn, yupToFormErrors, validateYupSchema, getIn} from './utils';
import set from 'lodash.set';

export default class FormViewModel {
  validate;
  validationSchema;
  initialValues = {};
  @observable values = {...this.initialValues};
  @observable touched = {};
  @observable errors = {};
  @observable isSubmitting = false;
  @observable submitCount = 0;

  getValueFromEvent(e) {
    const {type, value, checked} = e.target;
    const isNumericalValue = /number|range/.test(type);
    const isCheckbox = /checkbox/.test(type);


    if (isNumericalValue) {
      let parsed = parseFloat(value);
      return isNaN(parsed) ? '' : parsed;
    }

    if (isCheckbox) {
      return checked;
    }

    return value;
  }


  handleChange = (e) => {
    const target = e.target || e;
    const {name, id} = target;
    const fieldName = name || id;
    const fieldValue = this.getValueFromEvent(e);
    if (e.persist) {
      e.persist();
    }
    this.setFieldValue(fieldName, fieldValue);
  }

  handleBlur = (e) => {
    const target = e.target || e;
    const {name, id} = target;
    const fieldName = name || id;
    this.setFieldTouched(fieldName, true);
  }

  setFieldValue = (name, value) => {
    this.setInMobx('values', name, value);
    this.runValidations();
  };

  setValues = (values) => {
    this.values = values;
    this.runValidations();
  };

  setInMobx(prop, name, value) {
    const lastValue = getIn(this[prop], name);

    if (lastValue === undefined) {
      this[prop] = setIn(this[prop], name, value);
    } else {
      set(this[prop], name, value);
    }
  }

  setFieldTouched = (name, value) => {
    // if (typeof this.touched[name] === "undefined") {
    //   this.touched = setIn(this.touched, name, value);
    // } else {
    //   this.touched[name] = value;
    // }

    this.setInMobx('touched', name, value);
    


    
    this.runValidations();
  };

  setFieldError = (name, value) => {
    this.errors[name] = value;
  };

  setErrors = (errors) => {
    this.errors = errors;
  };

  resetForm = () => {
    this.values = {...this.initialValues};
    this.touched = {};
    this.errors = {};
    this.isSubmitting = false;
  };

  async runValidationSchema(values) {
    const schema = isFunction(this.validationSchema) ? this.validationSchema() : this.validationSchema;
    try {
      const results = await validateYupSchema(values, schema);
      return {};
    } catch(err) {
      const errors = yupToFormErrors(err);
      console.log(errors);
      this.isSubmitting = false;
      return errors;
    }
  }

  async runValidations(values = this.values) {
    let schemaErrors;
    let validateFnErrors;

    if (this.validationSchema) {
      schemaErrors = await this.runValidationSchema(values);
    }

    if (this.validate) {
      validateFnErrors = await this.validate(this.values);
    }

    this.errors = {...schemaErrors, ...validateFnErrors};
  }

  submitForm = async () => {
    this.touched = setNestedObjectValues(this.values, true);
    this.isSubmitting = true;
    this.submitCount = this.submitCount + 1;

    if (this.validate) {
      this.errors = await this.validate(this.values);
      const isValid = Object.keys(this.errors).length === 0;

      if (!isValid) {
        return;
      }
    }

    if (this.validationSchema) {
      this.errors = await this.runValidationSchema(this.values);
    }

    return this.executeSubmit();
  }

  executeSubmit() {

  }

  onReset(values?) {

  }

  handleSubmit = (e) => {
    if (e && e.preventDefault) {
      e.preventDefault();
    }

    return this.submitForm();
  }

  private handleReset = () => {
    if (this.onReset) {
      this.onReset(this.values);
    }
  };
}
