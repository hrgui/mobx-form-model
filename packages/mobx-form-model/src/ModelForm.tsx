import { observer, Provider, inject } from 'mobx-react';
import * as React from 'react';
import FormViewModel from './FormViewModel';
import {isFunction} from 'lodash';
import { observable } from 'mobx';

export interface ModelFormProps {
  modelConstructor?: any;
  initialValues?: any;
  values?: any;
  model?: FormViewModel;
  onSubmitSuccess?;
  onSubmitError?;
  validate?;
  parentModel?: FormViewModel;
  name?: string;
  children?: any;
  validationSchema?;
  onSubmit?;
}

@inject((stores: any) => {
  return {parentModel: stores.model}
})
@observer
export default class ModelForm extends React.Component<ModelFormProps, any> {
  @observable model;

  generateModel() {
    const ModelConstructor = this.props.modelConstructor || FormViewModel;
    this.model = this.props.model || new ModelConstructor(this.props.initialValues);

    if (this.props.parentModel && this.props.name) {
      const parentModel = this.props.parentModel;
      parentModel.addChildFormModel(this.model, this.props.name);
    }
  }

  updateModel(props, prevProps: any = {}) {
    let initModel = false;
    if (!this.model) {
      this.generateModel();
      initModel = true;
    }


    const {onSubmitSuccess, onSubmit, validationSchema, onSubmitError} = props;
    const {model} = this;

    if (!initModel && prevProps.initialValues !== props.initialValues) {
      model.setValues(props.initialValues);
    }

    if (!initModel && prevProps.values !== props.values) {
      model.setValues(props.values);
    }

    if (onSubmitSuccess) {
      model.onSubmitSuccess = onSubmitSuccess.bind(model);
    }

    if (onSubmitError) {
      model.onSubmitError = onSubmitError.bind(model);
    }

    if (onSubmit) {
      model.onSubmit = onSubmit.bind(model);
    }

    if (validationSchema) {
      model.validationSchema = validationSchema;
    }
  }

  componentDidMount() {
    this.updateModel(this.props);
  }


  componentDidUpdate(prevProps) {
    this.updateModel(this.props, prevProps);
  }

  render() {
    const {children} = this.props;
    const {model} = this;

    if (!model) {
      return null;
    }

    return (
      <Provider model={model}>
        <React.Fragment>
          {isFunction(children) ? children({model}) : children}
        </React.Fragment>
      </Provider>
    );
  }
}
