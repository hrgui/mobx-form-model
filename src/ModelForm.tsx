import {observer} from 'mobx-react';
import * as React from 'react';
import FormViewModel from './FormViewModel';
export const ModelFormContext = React.createContext<FormViewModel>(null);

@observer
export default class ModelForm extends React.Component<any, any> {
  render() {
    return <ModelFormContext.Provider value={this.props.model}>
      {this.props.children}
    </ModelFormContext.Provider>
  }
};