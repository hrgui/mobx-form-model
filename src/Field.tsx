import { ModelFormContext } from './ModelForm';
import * as React from 'react';
import { Observer, observer } from 'mobx-react';
import { getIn, isFunction } from './utils';
import filterReactProps from 'filter-react-props';


@observer
export default class Field extends React.Component<any, any> {
  rendercount = 0;
  render() {
    return <ModelFormContext.Consumer>
      {(model) => {
        if (!model) {
          return null;
        }
        return <Observer>
          {() => {
            const { type, name, value, component = 'input', render } = this.props;
            const modelValue = getIn(model.values, name);
            const defaultProps = {
              // why OR '' so we dont get uncontrolled error
              value: (type === 'radio' || type === 'checkbox' ? value : modelValue) || '',
              onChange: model.handleChange,
              onBlur: model.handleBlur,
              checked: type === 'radio' ? modelValue === this.props.value : undefined
            };

            const props = {...defaultProps, ...this.props, model, _rendercount: this.rendercount};

            this.rendercount += 1;

            if (isFunction(render)) {
              return render(props);
            }

            

            return React.createElement(component, {...filterReactProps(props), _rendercount: this.rendercount});
          }}
        </Observer>
      }}
    </ModelFormContext.Consumer>
  }
};