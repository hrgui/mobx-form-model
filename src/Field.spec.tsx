import * as React from 'react';
import * as renderer from 'react-test-renderer';
import Field from './Field';
import ModelForm from './ModelForm';
import FormViewModel from './FormViewModel';

describe('Field', () => {
  it('should render null w/o ModelForm', () => {
    const tree = renderer.create(<Field />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  describe('Field <=> ModelForm', () => {
    it('should render basic types easily (text, number, radio)', () => {
      const formViewModel = new FormViewModel();
      formViewModel.setFieldValue('name', 'Mickey Mouse');
      formViewModel.setFieldValue('age', 100);
      formViewModel.setFieldValue('gender', 'male');
      formViewModel.setFieldValue('isCartoon', true);
      const tree = renderer
        .create(
          <ModelForm model={formViewModel}>
            <Field name="name" />
            <Field name="age" type="number" />
            <Field name="gender" value="male" type="radio" />
            <Field name="gender" value="female" type="radio" />
            <Field name="isCartoon" />
          </ModelForm>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render basic types easily with a custom type in addition', () => {
      const formViewModel = new FormViewModel();
      formViewModel.setFieldValue('name', 'Mickey Mouse');
      formViewModel.setFieldValue('age', 100);
      formViewModel.setFieldValue('gender', 'male');
      formViewModel.setFieldValue('isCartoon', true);
      const ReadOnlyComponent = ({ value }) => {
        return value;
      };
      const tree = renderer
        .create(
          <ModelForm model={formViewModel}>
            <Field name="name" />
            <Field name="gender" component={ReadOnlyComponent} />
          </ModelForm>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should support nested values', () => {
      const formViewModel = new FormViewModel();
      formViewModel.setFieldValue('personal', {
        name: 'Mickey Mouse'
      });
      const tree = renderer
        .create(
          <ModelForm model={formViewModel}>
            <Field name="personal.name" />
          </ModelForm>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
