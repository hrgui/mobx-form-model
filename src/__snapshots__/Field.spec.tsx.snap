// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field Field <=> ModelForm should render basic types easily (text, number, radio) 1`] = `
Array [
  <input
    defaultValue="Mickey Mouse"
    name="name"
    onBlur={[Function]}
    onChange={[Function]}
    value="Mickey Mouse"
  />,
  <input
    defaultValue={100}
    name="age"
    onBlur={[Function]}
    onChange={[Function]}
    type="number"
    value={100}
  />,
  <input
    checked={true}
    defaultValue="male"
    name="gender"
    onBlur={[Function]}
    onChange={[Function]}
    type="radio"
    value="male"
  />,
  <input
    checked={false}
    defaultValue="female"
    name="gender"
    onBlur={[Function]}
    onChange={[Function]}
    type="radio"
    value="female"
  />,
  <input
    defaultValue={true}
    name="isCartoon"
    onBlur={[Function]}
    onChange={[Function]}
    value={true}
  />,
]
`;

exports[`Field Field <=> ModelForm should render basic types easily with a custom type in addition 1`] = `
Array [
  <input
    defaultValue="Mickey Mouse"
    name="name"
    onBlur={[Function]}
    onChange={[Function]}
    value="Mickey Mouse"
  />,
  "male",
]
`;

exports[`Field Field <=> ModelForm should support nested values 1`] = `
<input
  defaultValue="Mickey Mouse"
  name="personal.name"
  onBlur={[Function]}
  onChange={[Function]}
  value="Mickey Mouse"
/>
`;

exports[`Field should render null w/o ModelForm 1`] = `null`;
